### ISSUE TITLE
Linux Server Issues "configure: error: C compiler cannot create executables"

### ISSUE THREAD

**Issue by ghost** [mastercomfig/tf2-patches#619]
### Description
While going through Linux Build Instructions, encountered error: "configure: error: C compiler cannot create executables". Still troubleshooting, although it could be helpful if the documentation were updated a bit more clearly (once I get this fixed and working, I might make a shell script to automate the entire process).

Got project working on Windows (client and server), attempting to run a TC2 server on Linux for testing/gaming purposes. I have experience with running Linux TF2 Servers, so I'm fairly familiar with the general process. Has anyone else encountered this issue?

**Comment by thereal-bumface** [comment]
Also had segmentation fault using DepotDownloader, had to use SteamCMD in order to download TF2 properly.

**Comment by melvyn2** [comment]
check the `config.log`, upload it here.

**Comment by thereal-bumface** [comment]
checking whether to enable maintainer-specific portions of Makefiles... yes
checking build system type... x86_64-unknown-linux-gnu
checking host system type... x86_64-unknown-linux-gnu
checking target system type... x86_64-unknown-linux-gnu
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /usr/bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking for gcc... /home/trb/team-comtress-2-main/src/devtools/bin/linux/ccache gcc
checking whether the C compiler works... no
configure: error: in `/home/trb/team-comtress-2-main/src/thirdparty/protobuf-2.6.1':
configure: error: C compiler cannot create executables
See `config.log' for more details
trb@tc2:~/team-comtress-2-main$ cd src/thirdparty/protobuf-2.6.1/
trb@tc2:~/team-comtress-2-main/src/thirdparty/protobuf-2.6.1$ cat config.log
This file contains any messages produced by compilers while
running configure, to aid debugging if configure makes a mistake.

It was created by Protocol Buffers configure 2.6.1, which was
generated by GNU Autoconf 2.69.  Invocation command line was

  $ ./configure CFLAGS=-m32 -Wno-reserved-user-defined-literal -D_GLIBCXX_USE_CXX11_ABI=0 -w CXXFLAGS=-m32 -Wno-reserved-user-defined-literal -D_GLIBCXX_USE_CXX11_ABI=0 -w LDFLAGS=-m32

## --------- ##
## Platform. ##
## --------- ##

hostname = tc2
uname -m = x86_64
uname -r = 5.15.0-27-generic
uname -s = Linux
uname -v = #28-Ubuntu SMP Thu Apr 14 04:55:28 UTC 2022

/usr/bin/uname -p = x86_64
/bin/uname -X     = unknown

/bin/arch              = x86_64
/usr/bin/arch -k       = unknown
/usr/convex/getsysinfo = unknown
/usr/bin/hostinfo      = unknown
/bin/machine           = unknown
/usr/bin/oslevel       = unknown
/bin/universe          = unknown

PATH: /usr/local/sbin
PATH: /usr/local/bin
PATH: /usr/sbin
PATH: /usr/bin
PATH: /sbin
PATH: /bin
PATH: /usr/games
PATH: /snap/bin
PATH: /home/trb/.dotnet/


## ----------- ##
## Core tests. ##
## ----------- ##

configure:2483: checking whether to enable maintainer-specific portions of Makefiles
configure:2492: result: yes
configure:2553: checking build system type
configure:2567: result: x86_64-unknown-linux-gnu
configure:2587: checking host system type
configure:2600: result: x86_64-unknown-linux-gnu
configure:2620: checking target system type
configure:2633: result: x86_64-unknown-linux-gnu
configure:2676: checking for a BSD-compatible install
configure:2744: result: /usr/bin/install -c
configure:2755: checking whether build environment is sane
configure:2810: result: yes
configure:2961: checking for a thread-safe mkdir -p
configure:3000: result: /usr/bin/mkdir -p
configure:3007: checking for gawk
configure:3023: found /usr/bin/gawk
configure:3034: result: gawk
configure:3045: checking whether make sets $(MAKE)
configure:3067: result: yes
configure:3096: checking whether make supports nested variables
configure:3113: result: yes
configure:3306: checking for gcc
configure:3333: result: /home/trb/team-comtress-2-main/src/devtools/bin/linux/ccache gcc
configure:3562: checking for C compiler version
configure:3571: /home/trb/team-comtress-2-main/src/devtools/bin/linux/ccache gcc --version >&5
gcc (Ubuntu 11.2.0-19ubuntu1) 11.2.0
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

configure:3582: $? = 0
configure:3571: /home/trb/team-comtress-2-main/src/devtools/bin/linux/ccache gcc -v >&5
Using built-in specs.
COLLECT_GCC=/usr/bin/gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/11/lto-wrapper
OFFLOAD_TARGET_NAMES=nvptx-none:amdgcn-amdhsa
OFFLOAD_TARGET_DEFAULT=1
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 11.2.0-19ubuntu1' --with-bugurl=file:///usr/share/doc/gcc-11/README.Bugs --enable-languages=c,ada,c++,go,brig,d,fortran,objc,obj-c++,m2 --prefix=/usr --with-gcc-major-version-only --program-suffix=-11 --program-prefix=x86_64-linux-gnu- --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --enable-bootstrap --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-plugin --enable-default-pie --with-system-zlib --enable-libphobos-checking=release --with-target-system-zlib=auto --enable-objc-gc=auto --enable-multiarch --disable-werror --enable-cet --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-offload-targets=nvptx-none=/build/gcc-11-gBFGDP/gcc-11-11.2.0/debian/tmp-nvptx/usr,amdgcn-amdhsa=/build/gcc-11-gBFGDP/gcc-11-11.2.0/debian/tmp-gcn/usr --without-cuda-driver --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu --with-build-config=bootstrap-lto-lean --enable-link-serialization=2
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 11.2.0 (Ubuntu 11.2.0-19ubuntu1)
... rest of stderr output deleted ...
configure:3582: $? = 0
configure:3571: /home/trb/team-comtress-2-main/src/devtools/bin/linux/ccache gcc -V >&5
gcc: error: unrecognized command-line option '-V'
gcc: fatal error: no input files
compilation terminated.
configure:3582: $? = 1
configure:3571: /home/trb/team-comtress-2-main/src/devtools/bin/linux/ccache gcc -qversion >&5
gcc: error: unrecognized command-line option '-qversion'; did you mean '--version'?
gcc: fatal error: no input files
compilation terminated.
configure:3582: $? = 1
configure:3602: checking whether the C compiler works
configure:3624: /home/trb/team-comtress-2-main/src/devtools/bin/linux/ccache gcc -m32 -Wno-reserved-user-defined-literal -D_GLIBCXX_USE_CXX11_ABI=0 -w  -m32 conftest.c  >&5
/usr/bin/ld: cannot find Scrt1.o: No such file or directory
/usr/bin/ld: cannot find crti.o: No such file or directory
/usr/bin/ld: skipping incompatible /usr/lib/gcc/x86_64-linux-gnu/11/libgcc.a when searching for -lgcc
/usr/bin/ld: cannot find -lgcc: No such file or directory
/usr/bin/ld: skipping incompatible /usr/lib/gcc/x86_64-linux-gnu/11/libgcc.a when searching for -lgcc
/usr/bin/ld: cannot find -lgcc: No such file or directory
collect2: error: ld returned 1 exit status
configure:3628: $? = 1
configure:3666: result: no
configure: failed program was:
| /* confdefs.h */
| #define PACKAGE_NAME "Protocol Buffers"
| #define PACKAGE_TARNAME "protobuf"
| #define PACKAGE_VERSION "2.6.1"
| #define PACKAGE_STRING "Protocol Buffers 2.6.1"
| #define PACKAGE_BUGREPORT "protobuf@googlegroups.com"
| #define PACKAGE_URL ""
| #define PACKAGE "protobuf"
| #define VERSION "2.6.1"
| /* end confdefs.h.  */
|
| int
| main ()
| {
|
|   ;
|   return 0;
| }
configure:3671: error: in `/home/trb/team-comtress-2-main/src/thirdparty/protobuf-2.6.1':
configure:3673: error: C compiler cannot create executables
See `config.log' for more details

## ---------------- ##
## Cache variables. ##
## ---------------- ##

ac_cv_build=x86_64-unknown-linux-gnu
ac_cv_env_CCC_set=
ac_cv_env_CCC_value=
ac_cv_env_CC_set=set
ac_cv_env_CC_value='/home/trb/team-comtress-2-main/src/devtools/bin/linux/ccache gcc'
ac_cv_env_CFLAGS_set=set
ac_cv_env_CFLAGS_value='-m32 -Wno-reserved-user-defined-literal -D_GLIBCXX_USE_CXX11_ABI=0 -w'
ac_cv_env_CPPFLAGS_set=
ac_cv_env_CPPFLAGS_value=
ac_cv_env_CXXCPP_set=
ac_cv_env_CXXCPP_value=
ac_cv_env_CXXFLAGS_set=set
ac_cv_env_CXXFLAGS_value='-m32 -Wno-reserved-user-defined-literal -D_GLIBCXX_USE_CXX11_ABI=0 -w'
ac_cv_env_CXX_set=set
ac_cv_env_CXX_value='/home/trb/team-comtress-2-main/src/devtools/bin/linux/ccache g++'
ac_cv_env_LDFLAGS_set=set
ac_cv_env_LDFLAGS_value=-m32
ac_cv_env_LIBS_set=
ac_cv_env_LIBS_value=
ac_cv_env_build_alias_set=
ac_cv_env_build_alias_value=
ac_cv_env_host_alias_set=
ac_cv_env_host_alias_value=
ac_cv_env_target_alias_set=
ac_cv_env_target_alias_value=
ac_cv_host=x86_64-unknown-linux-gnu
ac_cv_path_install='/usr/bin/install -c'
ac_cv_path_mkdir=/usr/bin/mkdir
ac_cv_prog_AWK=gawk
ac_cv_prog_ac_ct_CC='/home/trb/team-comtress-2-main/src/devtools/bin/linux/ccache gcc'
ac_cv_prog_make_make_set=yes
ac_cv_target=x86_64-unknown-linux-gnu
am_cv_make_support_nested_variables=yes

## ----------------- ##
## Output variables. ##
## ----------------- ##

ACLOCAL='${SHELL} /home/trb/team-comtress-2-main/src/thirdparty/protobuf-2.6.1/missing aclocal-1.14'
AMDEPBACKSLASH=''
AMDEP_FALSE=''
AMDEP_TRUE=''
AMTAR='$${TAR-tar}'
AM_BACKSLASH='\'
AM_DEFAULT_V='$(AM_DEFAULT_VERBOSITY)'
AM_DEFAULT_VERBOSITY='1'
AM_V='$(V)'
AR=''
AUTOCONF='${SHELL} /home/trb/team-comtress-2-main/src/thirdparty/protobuf-2.6.1/missing autoconf'
AUTOHEADER='${SHELL} /home/trb/team-comtress-2-main/src/thirdparty/protobuf-2.6.1/missing autoheader'
AUTOMAKE='${SHELL} /home/trb/team-comtress-2-main/src/thirdparty/protobuf-2.6.1/missing automake-1.14'
AWK='gawk'
CC='/home/trb/team-comtress-2-main/src/devtools/bin/linux/ccache gcc'
CCDEPMODE=''
CFLAGS='-m32 -Wno-reserved-user-defined-literal -D_GLIBCXX_USE_CXX11_ABI=0 -w'
CPPFLAGS=''
CXX='/home/trb/team-comtress-2-main/src/devtools/bin/linux/ccache g++'
CXXCPP=''
CXXDEPMODE=''
CXXFLAGS='-m32 -Wno-reserved-user-defined-literal -D_GLIBCXX_USE_CXX11_ABI=0 -w'
CYGPATH_W='echo'
DEFS=''
DEPDIR=''
DLLTOOL=''
DSYMUTIL=''
DUMPBIN=''
ECHO_C=''
ECHO_N='-n'
ECHO_T=''
EGREP=''
EXEEXT=''
FGREP=''
GCC_FALSE=''
GCC_TRUE=''
GREP=''
HAVE_ZLIB_FALSE=''
HAVE_ZLIB_TRUE=''
INSTALL_DATA='${INSTALL} -m 644'
INSTALL_PROGRAM='${INSTALL}'
INSTALL_SCRIPT='${INSTALL}'
INSTALL_STRIP_PROGRAM='$(install_sh) -c -s'
ISAINFO=''
LD=''
LDFLAGS='-m32'
LIBOBJS=''
LIBS=''
LIBTOOL=''
LIPO=''
LN_S=''
LTLIBOBJS=''
MAINT=''
MAINTAINER_MODE_FALSE='#'
MAINTAINER_MODE_TRUE=''
MAKEINFO='${SHELL} /home/trb/team-comtress-2-main/src/thirdparty/protobuf-2.6.1/missing makeinfo'
MANIFEST_TOOL=''
MKDIR_P='/usr/bin/mkdir -p'
NM=''
NMEDIT=''
OBJDUMP=''
OBJEXT=''
OTOOL64=''
OTOOL=''
PACKAGE='protobuf'
PACKAGE_BUGREPORT='protobuf@googlegroups.com'
PACKAGE_NAME='Protocol Buffers'
PACKAGE_STRING='Protocol Buffers 2.6.1'
PACKAGE_TARNAME='protobuf'
PACKAGE_URL=''
PACKAGE_VERSION='2.6.1'
PATH_SEPARATOR=':'
POW_LIB=''
PROTOBUF_OPT_FLAG=''
PROTOC=''
PTHREAD_CC=''
PTHREAD_CFLAGS=''
PTHREAD_LIBS=''
RANLIB=''
SED=''
SET_MAKE=''
SHELL='/bin/bash'
STRIP=''
USE_EXTERNAL_PROTOC_FALSE=''
USE_EXTERNAL_PROTOC_TRUE=''
VERSION='2.6.1'
ac_ct_AR=''
ac_ct_CC='/home/trb/team-comtress-2-main/src/devtools/bin/linux/ccache gcc'
ac_ct_CXX=''
ac_ct_DUMPBIN=''
acx_pthread_config=''
am__EXEEXT_FALSE=''
am__EXEEXT_TRUE=''
am__fastdepCC_FALSE=''
am__fastdepCC_TRUE=''
am__fastdepCXX_FALSE=''
am__fastdepCXX_TRUE=''
am__include=''
am__isrc=''
am__leading_dot='.'
am__nodep=''
am__quote=''
am__tar='$${TAR-tar} chof - "$$tardir"'
am__untar='$${TAR-tar} xf -'
bindir='${exec_prefix}/bin'
build='x86_64-unknown-linux-gnu'
build_alias=''
build_cpu='x86_64'
build_os='linux-gnu'
build_vendor='unknown'
datadir='${datarootdir}'
datarootdir='${prefix}/share'
docdir='${datarootdir}/doc/${PACKAGE_TARNAME}'
dvidir='${docdir}'
exec_prefix='NONE'
host='x86_64-unknown-linux-gnu'
host_alias=''
host_cpu='x86_64'
host_os='linux-gnu'
host_vendor='unknown'
htmldir='${docdir}'
includedir='${prefix}/include'
infodir='${datarootdir}/info'
install_sh='${SHELL} /home/trb/team-comtress-2-main/src/thirdparty/protobuf-2.6.1/install-sh'
libdir='${exec_prefix}/lib'
libexecdir='${exec_prefix}/libexec'
localedir='${datarootdir}/locale'
localstatedir='${prefix}/var'
mandir='${datarootdir}/man'
mkdir_p='$(MKDIR_P)'
oldincludedir='/usr/include'
pdfdir='${docdir}'
prefix='NONE'
program_transform_name='s,x,x,'
psdir='${docdir}'
sbindir='${exec_prefix}/sbin'
sharedstatedir='${prefix}/com'
subdirs=''
sysconfdir='${prefix}/etc'
target='x86_64-unknown-linux-gnu'
target_alias=''
target_cpu='x86_64'
target_os='linux-gnu'
target_vendor='unknown'

## ----------- ##
## confdefs.h. ##
## ----------- ##

/* confdefs.h */
#define PACKAGE_NAME "Protocol Buffers"
#define PACKAGE_TARNAME "protobuf"
#define PACKAGE_VERSION "2.6.1"
#define PACKAGE_STRING "Protocol Buffers 2.6.1"
#define PACKAGE_BUGREPORT "protobuf@googlegroups.com"
#define PACKAGE_URL ""
#define PACKAGE "protobuf"
#define VERSION "2.6.1"

configure: exit 77

**Comment by melvyn2** [comment]
```

/usr/bin/ld: cannot find Scrt1.o: No such file or directory
/usr/bin/ld: cannot find crti.o: No such file or directory
/usr/bin/ld: skipping incompatible /usr/lib/gcc/x86_64-linux-gnu/11/libgcc.a when searching for -lgcc
/usr/bin/ld: cannot find -lgcc: No such file or directory
/usr/bin/ld: skipping incompatible /usr/lib/gcc/x86_64-linux-gnu/11/libgcc.a when searching for -lgcc
/usr/bin/ld: cannot find -lgcc: No such file or directory
```
You’re missing 32-bit libgcc, check your repos for it and I’ll add it to the dependencies if its missing.

**Comment by melvyn2** [comment]
I’ll take a look later today when I can

**Comment by thereal-bumface** [comment]
`../../public/tier1/fmtstr.h: In member function 'void CNumStr::SetUint64(uint64)':
../../public/tier1/fmtstr.h:333:51: error: 'Q_snprintf' was not declared in this scope; did you mean 'V_snprintf'?
  333 |         inline void SetUint64( uint64 un64 )    { Q_snprintf( m_szBuf, sizeof(m_szBuf), "%llu", un64 ); }
      |                                                   ^~~~~~~~~~
      |                                                   V_snprintf
make: *** [Makefile:267: obj/Linux/release/exprsimplifier.o] Error 1`

Seems like a duplicate of #615

**Comment by thereal-bumface** [comment]
Started server with -insecure as a start option. Now I get

`Disconnect: Server uses different class tables`

**Comment by thereal-bumface** [comment]
Ran
`login anonymous
force_install_dir /hlserver/tf2
app_update 232250 validate
quit`

Now the TC2 Linux server says it's running on a newer version of the game. I can connect via TF2, but not TC2.

CLIENT TC2 (Windows)
`] version
Build Label:           4139084   # Uniquely identifies each build
Network PatchVersion:  4085139   # Determines client and server compatibility
Protocol version:           24   # High level network protocol version
Server version:        4139084
Server AppID:           232250
Client version:        4085139
Client AppID:              440`

SERVER TC2 (Linux)
`version
Build Label:           7182415   # Uniquely identifies each build
Network PatchVersion:  7182415   # Determines client and server compatibility
Protocol version:           24   # High level network protocol version
Server version:        7182415
Server AppID:           232250`

**Comment by thereal-bumface** [comment]
Hrm, so I unzipped game_clean.zip again, and the versions of the client/server match, however I'm back to the error I see client side when attempting to connect to the server:
`Disconnect: Server uses different class tables`

**Comment by thereal-bumface** [comment]
Will do; I want to see if I can get it working in the first place; looks like some of the symbolic links aren't created properly (extra forward slashes, etc.). I'll let you know once I've gotten something working

**Comment by thereal-bumface** [comment]
Still haven't managed to get it working; I've tried SteamCMD, DepotDownloader, I've tried copying the files from the linux box to my windows pc, and I can't manage to get a Linux Server Windows Client working without the `server uses different class tables` error. Not sure what could possibly be the issue, since I literally had identical files on both systems. 

Basic Process:
1. Download TF2 on the linux box via SteamCMD or DepotDownloader (after downloading dotnet requirements/packages)
2. Unzip game_clean.zip
3. ./build.sh is still getting `did you mean 'V_snprintf'?` issues
4. ./link.sh is still having issues with certain files
5. When I run tf2.sh (https://wiki.teamfortress.com/wiki/Linux_dedicated_server) (Create a shell script to run the server), it runs, but still class tables mismatch.

I'm assuming the class tables issue is being caused by the V_snprintf or issues linking files. I'll see if I can fix that issue.

**Comment by melvyn2** [comment]
oh, I see what happened: you didn’t build the server all. Rather, link.sh is linking in plain tf2, as there were no tc2 files already in the game dir.

**Comment by thereal-bumface** [comment]
Gotcha, I see your fixes; I guess they weren't added to main? So far just changing Q_snprintf to V_snprintf in fmtstr.h. I'll update on progress. Thanks for the quick response!

**Comment by melvyn2** [comment]
You’ll see why it’s still a draft soon enough (or just read the issue :p)

> Gotcha, I see your fixes; I guess they weren't added to main? So far just changing Q_snprintf to V_snprintf in fmtstr.h. I'll update on progress. Thanks for the quick response!

**Comment by thereal-bumface** [comment]
Hrm, so I should expect a segfault error on launch. Well glad I can help a bit on testing this out. It's still chugging along building, so hopefully it works out and I can help w/documentation.

**Comment by thereal-bumface** [comment]
Hrm; the Digital Ocean droplet I'm working with has a single core, and it looks like the tutorial says `it took 20 minutes on a Ryzen 5 1600.`, which has 6 cores. I'm wondering if I should create a Linux VM locally and allocate extra cores to speed up the build process, then transfer over.

**Comment by thereal-bumface** [comment]
Had to build 3 times in a row, it kept catching things. Got this error: 
`/home/trb/team-comtress-2-main/src/dedicated/console/TextConsoleUnix.cpp:16:10: fatal error: histedit.h: No such file or directory
   16 | #include <histedit.h>
      |          ^~~~~~~~~~~~
compilation terminated.
make[2]: *** [dedicated_linux32.mak:505: obj_dedicated_linux32/release/TextConsoleUnix.o] Error 1
make[1]: *** [games.mak:102: dedicated] Error 2
make: *** [games.mak:65: all] Error 2
real    0m58.570s
user    0m54.139s
sys     0m3.836s`

**Comment by thereal-bumface** [comment]
Fixed by using `sudo apt-get install libedit-dev`, apparently this resolves the issue of `histedit.h` missing.

**Comment by thereal-bumface** [comment]
Building: vtf
Building: client_tf
Building: datacache
Building: dedicated
---- LINKING obj_dedicated_linux32/release/dedicated.so [release] ----
/usr/bin/ld: skipping incompatible /usr/lib/x86_64-linux-gnu/libedit.so when searching for -ledit
/usr/bin/ld: skipping incompatible /usr/lib/x86_64-linux-gnu/libedit.a when searching for -ledit
/usr/bin/ld: cannot find -ledit: No such file or directory
/usr/bin/ld: skipping incompatible /usr/lib/x86_64-linux-gnu/libedit.so when searching for -ledit
collect2: error: ld returned 1 exit status
make[2]: *** [../devtools/makefile_base_posix.mak:583: obj_dedicated_linux32/release/dedicated.so] Error 1
make[1]: *** [games.mak:102: dedicated] Error 2
make: *** [games.mak:65: all] Error 2

real    0m1.008s
user    0m0.770s
sys     0m0.224s

**Comment by thereal-bumface** [comment]
Fixed with `sudo apt-get install libedit-dev:i386`.
Now I get
`---- LINKING EXE ../../game/srcds_linux [release] ----
/usr/bin/ld: cannot find -ltcmalloc_minimal: No such file or directory
collect2: error: ld returned 1 exit status
make[2]: *** [../devtools/makefile_base_posix.mak:595: ../../game/srcds_linux] Error 1
make[1]: *** [games.mak:106: dedicated_main] Error 2
make: *** [games.mak:65: all] Error 2
real    0m0.971s
user    0m0.695s
sys     0m0.262s`

**Comment by melvyn2** [comment]
Missing `libtcmalloc-minimal4:i386` 
Did you run the dependencies command?

**Comment by thereal-bumface** [comment]
Yeah, just ran it again, got same error as before: `E: Unable to locate package libtcmalloc-minimal4:i386`

**Comment by thereal-bumface** [comment]
Testing removing "-ltcmalloc_minimal -Wl" from the line in source_exe_posix_base.vpc completely
`$GCC_ExtraLinkerFlags			"-Wl,--no-as-needed -ltcmalloc_minimal -Wl,--as-needed" [$LINUXALL&&!$DEDICATED]`

**Comment by thereal-bumface** [comment]
Got another issue with `#include <AL/al.h>` not being found, fixed it by installing `sudo apt-get install libopenal1:i386 libopenal1`, manually as well from `https://packages.debian.org/stretch/i386/libopenal-dev/download`

**Comment by thereal-bumface** [comment]
Got another issue where libcurl gnutls was not found, installed via `sudo apt-get install -y libcurl4-gnutls-dev`, and `sudo apt --fix-broken`

**Comment by thereal-bumface** [comment]
Updated this line in engine.vpc `	$SystemLibraries				"rt;openal;curl-gnutls" [$LINUXALL]` to `	$SystemLibraries				"rt;openal;libcurl4-gnutls" [$LINUXALL]

**Comment by thereal-bumface** [comment]
Getting a strange related error: `/usr/bin/ld: cannot find -llibcurl4-gnutls: No such file or directory`. Not sure why there are 2 L's in the library now.

**Comment by melvyn2** [comment]
because you shouldn't have done that change. `-l<name>` means "link to the library named `lib<name>`

**Comment by melvyn2** [comment]
Ubuntu seems to have dropped a lot of needed `i386` packages. Personally I'd recommend switching to something not as breaking, like debian. As for `curl-gnutls`: it should be from the `libcurl4-gnutls-dev:i386` package

**Comment by melvyn2** [comment]
> Testing removing "-ltcmalloc_minimal -Wl" from the line in source_exe_posix_base.vpc completely
This is a bad idea, internal libs depend on tcmalloc's specific behavior

**Comment by thereal-bumface** [comment]
Gotcha, I might try another version of linux if I can't get this working. I see what you mean with -l<name> as well. Yeah the i386 packages seem to be having more issues. I'll see about putting back -ltcmalloc_minimal.

**Comment by thereal-bumface** [comment]
ran `sudo apt install mesa-common-dev` for missing gl.h

**Comment by thereal-bumface** [comment]
---- tf_gcmessages.pb.cc ----
{standard input}: Assembler messages:
{standard input}:2441168: Error: unknown pseudo-op: `.lvus493'
g++: fatal error: Killed signal terminated program cc1plus
compilation terminated.
make[2]: *** [server_linux32_tf.mak:9644: obj_server_tf_linux32/release/tf_gcmessages.pb.o] Error 1
make[1]: *** [games.mak:178: server_tf] Error 2
make: *** [games.mak:65: all] Error 2

real    1m2.320s
user    0m40.070s
sys     0m4.880s

**Comment by thereal-bumface** [comment]
Looks like I can make progress by building consecutively? It's not crashing where it did before, I seem to be making progress.

**Comment by melvyn2** [comment]
You can always keep building the working projects... ignoring the failing ones won't get you a working build. You'll just finish building the working ones.

**Comment by thereal-bumface** [comment]
Hmm, seems like it's getting stuck on this now:

Building: scenefilecache
Building: server_tf
-P-- tf_gamerules.cpp ----

Doesn't seem to be crashing/exiting/stopping, although I can do so manually with Ctrl+C. I'll wait it out for a bit.

**Comment by thereal-bumface** [comment]
`Writing master makefile /home/trb/team-comtress-2-main/src/games.mak.
virtual void CMaterialSystem::EndFrame()':
/home/trb/team-comtress-2-main/src/materialsystem/cmaterialsystem.cpp:3815:42: error: 'class CThreadMutex' has no member named 'GetOwnerId'`

**Comment by melvyn2** [comment]
check the fixes pr again...

**Comment by thereal-bumface** [comment]
I'm going to just download your for-pr entirely replace what I have, that way I don't have to look around for cherry picks.

**Comment by thereal-bumface** [comment]
Finally finished using your `for-pr`, got following:

---- sys_linux.cpp ----
---- LINKING obj_dedicated_linux32/debug/dedicated.so [debug] ----
/usr/bin/ld: ./obj_dedicated_linux32/debug/zip_utils.o: in function `CWin32File::CreateTempFile(CUtlString&, CUtlString&)':
/home/trb/team-comtress-2-main/src/public/zip_utils.cpp:181: warning: the use of `tmpnam' is dangerous, better use `mkstemp'
/usr/bin/ld: ./obj_dedicated_linux32/debug/TextConsoleUnix.o: warning: relocation against `_ZN23__pthread_cleanup_classD1Ev' in read-only section `.text'
/usr/bin/ld: warning: creating DT_TEXTREL in a shared object
---- COPYING TO ../../game/bin/dedicated.so [debug] ----
Building: dedicated_main
---- SteamAppStartup.cpp ----
---- main.cpp ----
---- LINKING EXE ../../game/srcds_linux [debug] ----
/usr/bin/ld: skipping incompatible /usr/lib/x86_64-linux-gnu/libtcmalloc_minimal.so when searching for -ltcmalloc_minimal
/usr/bin/ld: skipping incompatible /usr/lib/x86_64-linux-gnu/libtcmalloc_minimal.a when searching for -ltcmalloc_minimal
/usr/bin/ld: cannot find -ltcmalloc_minimal: No such file or directory
/usr/bin/ld: skipping incompatible /usr/lib/x86_64-linux-gnu/libtcmalloc_minimal.so when searching for -ltcmalloc_minimal
collect2: error: ld returned 1 exit status
make[2]: *** [../devtools/makefile_base_posix.mak:595: ../../game/srcds_linux] Error 1
make[1]: *** [games.mak:106: dedicated_main] Error 2
make: *** [games.mak:65: all] Error 2

**Comment by thereal-bumface** [comment]
Similar after trying again:

-P-- weapons_resource.cpp ----
-P-- weapon_selection.cpp ----
-P-- WorldDimsProxy.cpp ----
---- LINKING obj_client_tf_linux32/release/client.so [release] ----
collect2: fatal error: ld terminated with signal 9 [Killed]
compilation terminated.
make[2]: *** [../../devtools/makefile_base_posix.mak:583: obj_client_tf_linux32/release/client.so] Error 1
make[1]: *** [games.mak:94: client_tf] Error 2
make: *** [games.mak:65: all] Error 2

real    60m7.444s
user    52m43.418s
sys     6m34.182s

**Comment by thereal-bumface** [comment]
I think my system may be running out of memory, I'll see if I can fix this.

**Comment by thereal-bumface** [comment]
Increasing swap size to 2GB (up from 0) to see if this changes anything.

**Comment by thereal-bumface** [comment]
Looks like the same error as before:

---- LINKING EXE ../../game/srcds_linux [release] ----
/usr/bin/ld: skipping incompatible /usr/lib/x86_64-linux-gnu/libtcmalloc_minimal.so when searching for -ltcmalloc_minimal /usr/bin/ld: skipping incompatible /usr/lib/x86_64-linux-gnu/libtcmalloc_minimal.a when searching for -ltcmalloc_minimal
/usr/bin/ld: cannot find -ltcmalloc_minimal: No such file or directory
/usr/bin/ld: skipping incompatible /usr/lib/x86_64-linux-gnu/libtcmalloc_minimal.so when searching for -ltcmalloc_minimal collect2: error: ld returned 1 exit status
make[2]: *** [../devtools/makefile_base_posix.mak:595: ../../game/srcds_linux] Error 1
make[1]: *** [games.mak:106: dedicated_main] Error 2
make: *** [games.mak:65: all] Error 2

real    0m0.934s
user    0m0.677s
sys     0m0.240s

**Comment by thereal-bumface** [comment]
`fatal error: ft2build.h: No such file or directory` fixed with `sudo apt-get install libfreetype6-dev`

**Comment by thereal-bumface** [comment]
`fatal error: fontconfig/fontconfig.h: No such file or directory` fixed with `sudo apt-get install libfontconfig1-dev`

**Comment by thereal-bumface** [comment]
Building: vguimatsurface
---- LINKING obj_vguimatsurface_linux32/release/vguimatsurface.so [release] ----
/usr/bin/ld: skipping incompatible /usr/lib/x86_64-linux-gnu/libfontconfig.so when searching for -lfontconfig
/usr/bin/ld: skipping incompatible /usr/lib/x86_64-linux-gnu/libfontconfig.a when searching for -lfontconfig
/usr/bin/ld: cannot find -lfontconfig: No such file or directory
/usr/bin/ld: skipping incompatible /usr/lib/x86_64-linux-gnu/libfontconfig.so when searching for -lfontconfig
/usr/bin/ld: skipping incompatible /usr/lib/x86_64-linux-gnu/libfreetype.so when searching for -lfreetype
/usr/bin/ld: skipping incompatible /usr/lib/x86_64-linux-gnu/libfreetype.a when searching for -lfreetype
/usr/bin/ld: cannot find -lfreetype: No such file or directory
/usr/bin/ld: skipping incompatible /usr/lib/x86_64-linux-gnu/libfreetype.so when searching for -lfreetype
collect2: error: ld returned 1 exit status
make[2]: *** [../devtools/makefile_base_posix.mak:583: obj_vguimatsurface_linux32/release/vguimatsurface.so] Error 1
make[1]: *** [games.mak:250: vguimatsurface] Error 2
make: *** [games.mak:65: all] Error 2

real    0m2.675s
user    0m1.970s
sys     0m0.671s

**Comment by thereal-bumface** [comment]
Combination of the following fixed it, looks good to run, will test:

Manually downloading / sudo dpkg -i from this url: http://mirrors.kernel.org/ubuntu/pool/main/f/fontconfig/fontconfig_2.13.1-4.2ubuntu5_i386.deb
sudo apt install libfontconfig1-dev
sudo apt --fix-broken

**Comment by thereal-bumface** [comment]
trb@tc2:~/team-comtress-2-main$ ./link.sh /home/trb/game/
fatal: not a git repository (or any of the parent directories): .git
ln: failed to create symbolic link './game/hl2': File exists
ln: failed to create symbolic link './game/platform': File exists
ln: failed to create symbolic link './game/tf/maps': File exists
mkdir: cannot create directory './game/tf/materials': File exists
ln: failed to create symbolic link './game/tf/materials/models': File exists
ln: failed to create symbolic link './game/tf/materials/vgui': File exists
ln: failed to create symbolic link './game/tf/media': File exists
ln: failed to create symbolic link './game/tf/resource': File exists
ln: failed to create symbolic link './game/tf/scripts': File exists
ln: failed to create symbolic link './game/tf/tf2_misc_000.vpk': File exists
ln: failed to create symbolic link './game/tf/tf2_misc_001.vpk': File exists
ln: failed to create symbolic link './game/tf/tf2_misc_002.vpk': File exists
ln: failed to create symbolic link './game/tf/tf2_misc_003.vpk': File exists
ln: failed to create symbolic link './game/tf/tf2_misc_004.vpk': File exists
ln: failed to create symbolic link './game/tf/tf2_misc_005.vpk': File exists
ln: failed to create symbolic link './game/tf/tf2_misc_006.vpk': File exists
ln: failed to create symbolic link './game/tf/tf2_misc_007.vpk': File exists
ln: failed to create symbolic link './game/tf/tf2_misc_008.vpk': File exists
ln: failed to create symbolic link './game/tf/tf2_misc_009.vpk': File exists
ln: failed to create symbolic link './game/tf/tf2_misc_010.vpk': File exists
ln: failed to create symbolic link './game/tf/tf2_misc_011.vpk': File exists
ln: failed to create symbolic link './game/tf/tf2_misc_012.vpk': File exists
ln: failed to create symbolic link './game/tf/tf2_misc_013.vpk': File exists
ln: failed to create symbolic link './game/tf/tf2_misc_014.vpk': File exists
ln: failed to create symbolic link './game/tf/tf2_misc_015.vpk': File exists
ln: failed to create symbolic link './game/tf/tf2_misc_016.vpk': File exists
ln: failed to create symbolic link './game/tf/tf2_misc_017.vpk': File exists
ln: failed to create symbolic link './game/tf/tf2_misc_018.vpk': File exists
ln: failed to create symbolic link './game/tf/tf2_misc_019.vpk': File exists
ln: failed to create symbolic link './game/tf/tf2_misc_020.vpk': File exists
ln: failed to create symbolic link './game/tf/tf2_misc_021.vpk': File exists
ln: failed to create symbolic link './game/tf/tf2_misc_022.vpk': File exists
ln: failed to create symbolic link './game/tf/tf2_misc_023.vpk': File exists
ln: failed to create symbolic link './game/tf/tf2_misc_024.vpk': File exists
ln: failed to create symbolic link './game/tf/tf2_misc_dir.vpk': File exists
ln: failed to create symbolic link './game/tf/tf2_sound_misc_dir.vpk': File exists
ln: failed to create symbolic link './game/tf/tf2_sound_vo_english_dir.vpk': File exists
ln: failed to create symbolic link './game/tf/tf2_textures_dir.vpk': File exists
ln: failed to create symbolic link './game/tf/modelsounds.cache': File exists
ln: failed to create symbolic link './game/tf/tf2_sound_misc.vpk.sound.cache': File exists
ln: failed to create symbolic link './game/tf/tf2_sound_vo_english.vpk.sound.cache': File exists
cp: cannot stat '/home/trb/game//tf/gamestate.txt': No such file or directory
cp: cannot stat '/home/trb/game//tf/glshaders.cfg': No such file or directory
cp: cannot stat '/home/trb/game//tf/videoconfig_linux.cfg': No such file or directory

**Comment by thereal-bumface** [comment]
trb@tc2:~/team-comtress-2-main$ ./run.sh
error: XDG_RUNTIME_DIR not set in the environment.

**Comment by thereal-bumface** [comment]
Do I need two copies of TF2? One generated from the ./build.sh, and one to link to using ./link.sh? Not sure my 25 GB storage on the Digital Ocean droplet is sufficient.

**Comment by thereal-bumface** [comment]
I think I can get around this by downloading the linux version using depotdownloader onto my windows gaming pc, then remove unnecessary large files such as unused maps, then go through the linking process again. 

To remove all symlinks: `find -type l -delete`

**Comment by thereal-bumface** [comment]
decided to pay for extra 20GB block storage, installing tf2 there, then I'll use link.sh after.

**Comment by melvyn2** [comment]
Personally I'd delete the `game` folder, rebuild, and re-link. I should add a check to link.sh to see if tc2 has been built, but meanwhile it should only ever be run after building fully. Yes, you need the stock tf2 assets to run it. However, you don't need two copies of the game assets, it's just symlinks.

**Comment by thereal-bumface** [comment]
I guess what I meant was the team-comtress-main folder was about 15-20 GB by itself, and TF2 would be another 20 GB. I'll try what you mentioned if I can't get it to work with block storage.

**Comment by thereal-bumface** [comment]
Just ran ./build.sh again

Building: dedicated
make[2]: *** No rule to make target '../lib/public/linux32/libSDL2.so', needed by 'obj_dedicated_linux32/release/dedicated.so'.  Stop.
make[1]: *** [games.mak:102: dedicated] Error 2
make: *** [games.mak:65: all] Error 2

real    0m0.882s
user    0m0.652s
sys     0m0.218s

**Comment by melvyn2** [comment]
https://github.com/mastercomfig/team-comtress-2/blob/main/src/lib/public/linux32/libSDL2.so
It should be there.

**Comment by thereal-bumface** [comment]
Where is your version of the team-comtress-2 main.zip? I just unzipped the official main.zip, and now I'm back at the beginning where we had issues with min vs. std::min.

**Comment by melvyn2** [comment]
you should definitely try cloning from the repo. Use a shallow clone if you're worried about space. In fact, I think your issue stems from the zips maybe not including symlinks.

**Comment by melvyn2** [comment]
Also, troubleshooting of this sort could be easier to finish on discord, if you have the choice: https://discord.gg/CuPb2zV

**Comment by thereal-bumface** [comment]
:) Trying to sign up right now, getting an infinite beep boop captcha.

**Comment by thereal-bumface** [comment]
Welp got a violation for TOS/Community guidelines, guess discord is a no-go.

**Comment by thereal-bumface** [comment]
I'll keep building your complete for-pr build, see if I make more progress. I'll try to keep commenting to a minimum.

**Comment by thereal-bumface** [comment]
Got the same error as before, not sure how to fix it: 

---- main.cpp ----
---- LINKING EXE ../../game/srcds_linux [release] ----
/usr/bin/ld: cannot find -ltcmalloc_minimal: No such file or directory
collect2: error: ld returned 1 exit status

**Comment by melvyn2** [comment]
You still didn't install it, only the x86_64 version.

**Comment by melvyn2** [comment]
again, I would recommend not using ubuntu, clearly it has dropped too many i386 packages to be reasonably usable

**Comment by thereal-bumface** [comment]
Gotcha, I'll try debian.

**Comment by thereal-bumface** [comment]
`bash: line 1: /bin/ar: No such file or directory`

**Comment by thereal-bumface** [comment]
also, I'm testing this in a docker container on windows, running Debian (this lets me test faster).

**Comment by melvyn2** [comment]
Did you run the dependency install commands? If you did, I know where the issue is, I'll commit a fix when I get home.

**Comment by thereal-bumface** [comment]
I get the following:
`E: Unable to locate package fontconfig:i386
E: Unable to locate package libcurl4-gnutls-dev:i386
E: Unable to locate package libedit2:i386
E: Unable to locate package libfontconfig1:i386
E: Unable to locate package libfontconfig1-dev:i386
E: Unable to locate package libfreetype-dev:i386
E: Unable to locate package libncurses-dev:i386
E: Unable to locate package libopenal-dev:i386
E: Unable to locate package libopenal1:i386
E: Unable to locate package libtcmalloc-minimal4:i386
E: Unable to locate package libunwind-dev:i386
E: Unable to locate package libx11-dev:i386
E: Unable to locate package mesa-common-dev:i386`

**Comment by thereal-bumface** [comment]
Manually installed the packages:

dpkg -i fontconfig_2.13.1-4.2_i386.deb
(Reading database ... 20926 files and directories currently installed.)
Preparing to unpack fontconfig_2.13.1-4.2_i386.deb ...
Unpacking fontconfig:i386 (2.13.1-4.2) over (2.13.1-4.2) ...
dpkg: dependency problems prevent configuration of fontconfig:i386:
 fontconfig:i386 depends on libc6 (>= 2.4).
 fontconfig:i386 depends on libfontconfig1 (>= 2.13.0).
 fontconfig:i386 depends on libfreetype6 (>= 2.8.1).

dpkg: error processing package fontconfig:i386 (--install):
 dependency problems - leaving unconfigured
Errors were encountered while processing:
 fontconfig:i386

**Comment by melvyn2** [comment]
You definitely should not install the packages manually, it's never going to work right that way. You did add the i386 arch and run `apt update`?

**Comment by thereal-bumface** [comment]
yup, I did `dpkg --add-architecture i386` then `apt update`.

**Comment by melvyn2** [comment]
Which debian version?

**Comment by thereal-bumface** [comment]
11.3

**Comment by melvyn2** [comment]
You've misconfigured something, those packages definitely exist: https://packages.debian.org/search?suite=all&arch=i386&searchon=names&keywords=mesa-common-dev
Check your /etc/apt/sources.list

**Comment by thereal-bumface** [comment]
# deb http://snapshot.debian.org/archive/debian/20220418T000000Z bullseye main
deb http://deb.debian.org/debian bullseye main
# deb http://snapshot.debian.org/archive/debian-security/20220418T000000Z bullseye-security main
deb http://security.debian.org/debian-security bullseye-security main
# deb http://snapshot.debian.org/archive/debian/20220418T000000Z bullseye-updates main
deb http://deb.debian.org/debian bullseye-updates main

**Comment by melvyn2** [comment]
remove the lines with the snapshot urls and apt update. I'll note that this is really past the scope of this issue tracker, these are wider issues you seem to be having. I'll try to still help for now.

**Comment by thereal-bumface** [comment]
Gotcha; looks like removing the commented lines in the sources.list allowed me to properly apt get all the prerequisites (learned something new!)

---- decompress.c ----
bash: line 1: /bin/ar: No such file or directory
make[2]: *** [../../devtools/makefile_base_posix.mak:544: lzma.a] Error 127
make[1]: *** [games.mak:142: lzma] Error 2
make[1]: *** Waiting for unfinished jobs....

---- halton.cpp ----
cp -f pch_tier0.h ./obj_tier0_linux32/release/pch_tier0.h
bash: line 1: /bin/ar: No such file or directory
make[2]: *** [../../devtools/makefile_base_posix.mak:544: bzip2.a] Error 127
make[1]: *** [games.mak:86: bzip2] Error 2

-P-- platform_posix.cpp ----
bash: line 1: /bin/ar: No such file or directory
make[2]: *** [../devtools/makefile_base_posix.mak:544: appframework.a] Error 127
make[1]: *** [games.mak:74: appframework] Error 2

-P-- memdbg.cpp ----
bash: line 1: /bin/ar: No such file or directory
make[2]: *** [../devtools/makefile_base_posix.mak:544: tier1.a] Error 127
make[1]: *** [games.mak:222: tier1] Error 2

-P-- minidump.cpp ----
---- LINKING obj_tier0_linux32/release/libtier0.so [release] ----
/usr/bin/ld: ./obj_tier0_linux32/release/vprof.o: warning: relocation against `_
ZN10CVProfNode9MarkFrameEv' in read-only section `.text'
/usr/bin/ld: warning: creating DT_TEXTREL in a shared object
---- COPYING TO ../../game/bin/libtier0.so [release] ----
../devtools/gendbg.sh: line 40: /bin/objcopy: No such file or directory
../devtools/gendbg.sh: line 41: /bin/objcopy: No such file or directory
---- COPYING TO IMPORT LIBRARY ../lib/public/linux32/libtier0.so ----
make: *** [games.mak:65: all] Error 2

real    0m0.659s
user    0m2.949s
sys     0m0.790s

**Comment by melvyn2** [comment]
pull from my branch, I committed a likely fix

**Comment by thereal-bumface** [comment]
gotcha, gimme a bit to download/unzip

**Comment by thereal-bumface** [comment]
Building: vstdlib
---- LINKING EXE ../../game/hl2_linux [release] ----
---- LINKING EXE ../../game/srcds_linux [release] ----
/usr/bin/ld: cannot find -ltcmalloc_minimal
collect2: error: ld returned 1 exit status
/usr/bin/ldmake[2]: *** [../devtools/makefile_base_posix.mak:595: ../../game/hl2_linux] Error 1
make[1]: *** [games.mak:138: launcher_main] Error 2
make[1]: *** Waiting for unfinished jobs....
: cannot find -ltcmalloc_minimal
collect2: error: ld returned 1 exit status
make[2]: *** [../devtools/makefile_base_posix.mak:595: ../../game/srcds_linux] Error 1
make[1]: *** [games.mak:106: dedicated_main] Error 2
---- LINKING EXE ../../../game/simdtest [release] ----
/usr/bin/ld: cannot find -ltcmalloc_minimal
collect2: error: ld returned 1 exit status
make[2]: *** [../../devtools/makefile_base_posix.mak:595: ../../../game/simdtest] Error 1
make[1]: *** [games.mak:194: simdtest] Error 2
make: *** [games.mak:65: all] Error 2

real    0m0.131s
user    0m0.567s
sys     0m0.156s

**Comment by thereal-bumface** [comment]
apt-get install libtcmalloc-minimal4:i386
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
libtcmalloc-minimal4:i386 is already the newest version (2.8.1-1).
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.

**Comment by melvyn2** [comment]
hmm, try `sudo ln -s /usr/lib/i386-linux-gnu/libtcmalloc_minimal.so.4 /usr/lib/i386-linux-gnu/libtcmalloc_minimal.so`

**Comment by thereal-bumface** [comment]
Nice, seems to be progressing more. Got histedit.h missing, fixed again with `apt-get install libedit-dev`.

**Comment by thereal-bumface** [comment]
Got the following:
---- memoverride.cpp ----
---- memoverride.cpp ----
/usr/bin/ld: skipping incompatible /usr/lib/x86_64-linux-gnu/libedit.so when searching for -ledit
/usr/bin/ld: skipping incompatible /usr/lib/x86_64-linux-gnu/libedit.a when searching for -ledit
/usr/bin/ld: cannot find -ledit
/usr/bin/ld: skipping incompatible /usr/lib/x86_64-linux-gnu/libedit.so when searching for -ledit
collect2: error: ld returned 1 exit status
make[2]: *** [../devtools/makefile_base_posix.mak:583: obj_dedicated_linux32/release/dedicated.so] Error 1
make[1]: *** [games.mak:102: dedicated] Error 2
make[1]: *** Waiting for unfinished jobs....

**Comment by thereal-bumface** [comment]
apt install libedit-dev
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
libedit-dev is already the newest version (3.1-20191231-2+b1).
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.

**Comment by melvyn2** [comment]
try `libedit-dev:i386`

**Comment by thereal-bumface** [comment]
Hey, that looks like it worked! Looks like it 100% built.

./link.sh /home/trb/TC2
./link.sh: line 33: git: command not found
cp: cannot stat '/home/trb/TC2/tf/gamestate.txt': No such file or directory
cp: cannot stat '/home/trb/TC2/tf/glshaders.cfg': No such file or directory
cp: cannot stat '/home/trb/TC2/tf/videoconfig_linux.cfg': No such file or directory

**Comment by melvyn2** [comment]
first, you need to clone with git, as I told you...

**Comment by thereal-bumface** [comment]
Gotcha, sorry been really busy today so I was probably distracted; just git cloned, got the following:

`autoreconf: 'configure.ac' or 'configure.in' is required`

**Comment by melvyn2** [comment]
https://github.com/melvyn2/team-comtress-2/blob/for-pr/src/thirdparty/protobuf-2.6.1/configure.ac
should be there

**Comment by melvyn2** [comment]
> second, if I change it to ./srcds_linux, I get Failed to open dedicated.so (dedicated.so: cannot open shared object file: No such file or directory)

this error is because `LD_LIBRARY_PATH` wasn't set. I'll work on adding the server binaries to `run.sh`, they're known not to work so they haven't been used much

**Comment by thereal-bumface** [comment]
/home/trb/team-comtress-2/src/thirdparty/protobuf-2.6.1# ls -l configure.ac
-rw-r--r-- 1 root root 5067 May  5 03:32 configure.ac

**Comment by melvyn2** [comment]
hmm, I'm able to get a full build and run the server to the DXMath segfault we get on the client. I'm not sure what the issue is on your end. What's `git log -n 5` tell you?

**Comment by thereal-bumface** [comment]
git log -n 5
commit 689e4d7ebdeeb42da997ca76a23adfd8b974a7ba (HEAD -> main, origin/main, origin/HEAD)
Author: xero-lib <jonathankurtz@protonmail.com>
Date:   Sun Feb 6 03:29:09 2022 -0600

    Fixed build.sh

commit f7c8bbe65e90277dd5e659687263229d1a05c215
Author: xero-lib <jonathankurtz@protonmail.com>
Date:   Sun Feb 6 02:14:38 2022 -0600

    Fixed build.sh

commit b6b2dcd852b1c8a24e5ebc1c1c374fdc4be977ed
Author: melvyn2 <melvyn2@brcok.tk>
Date:   Sat Feb 5 19:16:01 2022 -0800

    Comment broken assert

commit 62f9f49d0992a211fd2a8e2c54a2a37fa1083feb
Author: melvyn2 <melvyn2@brcok.tk>
Date:   Sat Feb 5 19:14:32 2022 -0800

    Ignore vpc generated project files

commit ac7fca6b01a8b972f1e3878ef9be12cf6c1d78d3
Author: melvyn2 <melvyn2@brcok.tk>
Date:   Sat Feb 5 17:33:59 2022 -0800

    Run autoreconf on protobuf source

I ran into segfault errors before, but I believe it was a RAM issue which I solved with a 2 GB swap file. Not sure if it's related.

**Comment by melvyn2** [comment]
No, I mean a segfault in the built binary. Try running `git reset --hard origin/for-pr`

**Comment by thereal-bumface** [comment]
Thanks, it's looking good. This is my first time contributing / working on someone else's Git project, so I'm not that familiar with grabbing specific branches or spokes of branches.

**Comment by thereal-bumface** [comment]
Built correctly, testing link and run

**Comment by thereal-bumface** [comment]
I'm wondering if it'd be possible to use wine or something to run the windows TC2 server on a linux box.

**Comment by thereal-bumface** [comment]
> Built correctly, testing link.sh and run.sh:
> 
> ./run.sh WARNING: setlocale('en_US.UTF-8') failed, using locale:'C'. International characters may not work.
> 
> Not sure how to run the server now that it's built and linked. I'm guessing run.sh just runs the client.

Fixed this by uncommenting out the en_US.UTF-8 line in /etc/locale.gen, then running locale-gen

**Comment by thereal-bumface** [comment]
I think I'm running into that segmentation fault now (had to run with -s tag):

./run.sh -s
./run.sh: line 29:   736 Segmentation fault      "$(pwd)/srcds_linux" ${ARGS} ${@:2}

**Comment by thereal-bumface** [comment]
./run.sh -d
(lldb) target create "/home/trb/team-comtress-2/game/hl2_linux"
Current executable set to '/home/trb/team-comtress-2/game/hl2_linux' (i386).
(lldb) settings set -- target.run-args  "-allowdebug" "-steam" "-game" "tf" "-insecure" "-novid" "-nojoy" "-nosteamcontroller" "-nohltv" "-particles" "1" "-noborder" "-particle_fallback" "2" "-dev" "-nobreakpad" "-console"
(lldb)

**Comment by melvyn2** [comment]
`./run.sh -s -d`

**Comment by thereal-bumface** [comment]
./run.sh -s -d
(lldb) target create "/home/trb/team-comtress-2/game/srcds_linux"
Current executable set to '/home/trb/team-comtress-2/game/srcds_linux' (i386).
(lldb) settings set -- target.run-args  "-allowdebug" "-steam" "-game" "tf" "-insecure" "-novid" "-nojoy" "-nosteamcontroller" "-nohltv" "-particles" "1" "-noborder" "-particle_fallback" "2" "-dev" "-nobreakpad" "-console"
(lldb) r
error: 'A' packet returned an error: 8
(lldb)

**Comment by melvyn2** [comment]
https://stackoverflow.com/questions/70108775/gdb-exits-immediately-process-finished-with-exit-code-1-or-lldb-a-packet-ret/70108776#70108776

**Comment by thereal-bumface** [comment]
Not sure where to put this; I run docker desktop on windows, run my container, then in the console I run `docker exec -it container_name bash`. Where would I put this exactly?

**Comment by thereal-bumface** [comment]
Actually, I'm going to see if I can transfer the files now that they're built properly to a debian VPS on digital ocean.

**Comment by thereal-bumface** [comment]
trb@tc2:~/team-comtress-2$ ./run.sh -s -d
(lldb) target create "/home/trb/team-comtress-2/game/srcds_linux"
Current executable set to '/home/trb/team-comtress-2/game/srcds_linux' (i386).
(lldb) settings set -- target.run-args  "-allowdebug" "-steam" "-game" "tf" "-insecure" "-novid" "-nojoy" "-nosteamcontroller" "-nohltv" "-particles" "1" "-noborder" "-particle_fallback" "2" "-dev" "-nobreakpad" "-console"
(lldb) r
Process 31275 launched: '/home/trb/team-comtress-2/game/srcds_linux' (i386)

WARNING: Failed to load 32-bit libtinfo.so.5 or libncurses.so.5.
  Please install (lib32tinfo5 / ncurses-libs.i686 / equivalent) to enable readline.

Process 31275 stopped
* thread #1, name = 'srcds_linux', stop reason = signal SIGSEGV: invalid address (fault address: 0x0)
    frame #0: 0xf4e6a50e materialsystem.so`_GLOBAL__sub_I_mat_texture_tracking at vector4d.h:645:17
   642  {
   643  #if !defined( _X360 )
   644  #ifdef USE_DXMATH
-> 645          this->AsM128() = DirectX::XMVectorReplicateInt( 0 );
   646  #else
   647          this->AsM128() = _mm_set1_ps( 0 );
   648  #endif
(lldb) quit
Quitting LLDB will kill one or more processes. Do you really want to proceed: [Y/n] y

**Comment by thereal-bumface** [comment]
I've attempted replacing vector4d.h with various version, but all have resulted in similar signal SIGSEGV:

https://raw.githubusercontent.com/mastercomfig/team-comtress-2/main/src/public/mathlib/vector4d.h
https://raw.githubusercontent.com/ValveSoftware/source-sdk-2013/master/sp/src/public/mathlib/vector4d.h
https://swarm.workshop.perforce.com/files/guest/knut_wikstrom/ValveSDKCode/public/vector4d.h

**Comment by melvyn2** [comment]
yeah, from the little I've looked into it so far it's not a regression in DXmath itself

**Comment by thereal-bumface** [comment]
Referencing the Linux Build Guide:
`Bug reports (crashes in game) are appreciated! Please send a full backtrace by running with LLDB and crashing - run.sh -d will launch lldb with the game. Once in LLDB at the terminal, enter r at the (lldb) prompt to run the game, then reproduce the crash. The game should freeze instead of quitting and in your terminal, LLDB should say what the error was, and be at another prompt. type bt to get a backtrace, and put it in an issue on HurricanePootis's fork with how you got there.`

I ran the `bt` command in lldb after the SIGSEGV, and this is what I got:

(lldb) bt
* thread #1, name = 'srcds_linux', stop reason = signal SIGSEGV: invalid address (fault address: 0x0)
  * frame #0: 0xf4e6a50e materialsystem.so`_GLOBAL__sub_I_mat_texture_tracking at vector4d.h:645:17
    frame #1: 0xf7fe304c ld-2.31.so`___lldb_unnamed_symbol58$$ld-2.31.so + 140
    frame #2: 0xf7fe3152 ld-2.31.so`___lldb_unnamed_symbol59$$ld-2.31.so + 130
    frame #3: 0xf7fe6457 ld-2.31.so`___lldb_unnamed_symbol88$$ld-2.31.so + 23
    frame #4: 0xf7eab694 libc.so.6`_dl_catch_exception + 244
    frame #5: 0xf7fe6f1b ld-2.31.so`___lldb_unnamed_symbol93$$ld-2.31.so + 987
    frame #6: 0xf7eab639 libc.so.6`_dl_catch_exception + 153
    frame #7: 0xf7fe6796 ld-2.31.so`___lldb_unnamed_symbol91$$ld-2.31.so + 166
    frame #8: 0xf7f782c8 libdl.so.2`___lldb_unnamed_symbol4$$libdl.so.2 + 120
    frame #9: 0xf7eab639 libc.so.6`_dl_catch_exception + 153
    frame #10: 0xf7eab700 libc.so.6`_dl_catch_error + 48
    frame #11: 0xf7f78b11 libdl.so.2`___lldb_unnamed_symbol12$$libdl.so.2 + 113
    frame #12: 0xf7f78364 libdl.so.2`dlopen + 84
    frame #13: 0xf77a1cdb dedicated.so`Sys_LoadLibrary(char const*, Sys_Flags) + 139
    frame #14: 0xf77a1e27 dedicated.so`Sys_LoadModule(char const*, Sys_Flags) + 215
    frame #15: 0xf77696c5 dedicated.so`CBaseFileSystem::LoadModule(this=<unavailable>, pFileName=<unavailable>, pPathID=<unavailable>, bValidatedDllOnly=true) at basefilesystem.cpp:5055:23
    frame #16: 0xf788a99a dedicated.so`CSteamAppSystemGroup::LoadModuleDLL(char const*) + 106
    frame #17: 0xf788b660 dedicated.so`CAppSystemGroup::LoadModule(char const*) + 144
    frame #18: 0xf788bbcd dedicated.so`CAppSystemGroup::AddSystems(AppSystemInfo_t*) + 61
    frame #19: 0xf779e560 dedicated.so`CSys::LoadModules(CDedicatedAppSystemGroup*) [inlined] CDedicatedAppSystemGroup::AddSystems(pSystems=0xffff9fd8, this=0xffffa2a8) at dedicated.h:67:42
    frame #20: 0xf779e557 dedicated.so`CSys::LoadModules(this=0xf794707c, pAppSystemGroup=0xffffa2a8) at sys_linux.cpp:272
    frame #21: 0xf779d73a dedicated.so`CDedicatedAppSystemGroup::Create() [inlined] CDedicatedAppSystemGroup::Create(this=0xffffa2a8) at sys_ded.cpp:232:23
    frame #22: 0xf779d6f0 dedicated.so`CDedicatedAppSystemGroup::Create(this=0xffffa2a8) at sys_ded.cpp:216
    frame #23: 0xf788b356 dedicated.so`CAppSystemGroup::OnStartup() + 38
    frame #24: 0xf788b4b2 dedicated.so`CAppSystemGroup::Run() + 34
    frame #25: 0xf788da9b dedicated.so`CSteamApplication::Main() + 43
    frame #26: 0xf788b4c4 dedicated.so`CAppSystemGroup::Run() + 52
    frame #27: 0xf7751cc5 dedicated.so`main at sys_ded.cpp:529:29
    frame #28: 0xf779e60d dedicated.so`::DedicatedMain(argc=18, argv=0xffffd564) at sys_linux.cpp:297:13
    frame #29: 0x5655640f srcds_linux`main(argc=18, argv=0xffffd564) at main.cpp:226:22
    frame #30: 0xf7d8ae46 libc.so.6`__libc_start_main + 262
    frame #31: 0x56556591 srcds_linux`_start + 49

**Comment by melvyn2** [comment]
you should definitely use triple backtick code blocks

**Comment by thereal-bumface** [comment]
Gotcha, thanks for the tip (again sorry about my inexperience, hopefully it isn't too painful :))

**Comment by thereal-bumface** [comment]
Trying ./build.sh -dd, and using HurricanePootis's vector4d.h: https://raw.githubusercontent.com/HurricanePootis/team-comtress-2/independent/public/mathlib/vector4d.h

**Comment by melvyn2** [comment]
switching out the dxmath headers won't change anything, I explained that it's not a regression in dxmath itself

**Comment by thereal-bumface** [comment]
trb@tc2:~/team-comtress-2$ ./run.sh -s -d
(lldb) target create "/home/trb/team-comtress-2/game/srcds_linux"
Current executable set to '/home/trb/team-comtress-2/game/srcds_linux' (i386).
(lldb) settings set -- target.run-args  "-allowdebug" "-steam" "-game" "tf" "-insecure" "-novid" "-nojoy" "-nosteamcontroller" "-nohltv" "-particles" "1" "-noborder" "-particle_fallback" "2" "-dev" "-nobreakpad" "-console"
(lldb) r
Process 62744 launched: '/home/trb/team-comtress-2/game/srcds_linux' (i386)

WARNING: Failed to load 32-bit libtinfo.so.5 or libncurses.so.5.
  Please install (lib32tinfo5 / ncurses-libs.i686 / equivalent) to enable readline.

 failed to dlopen /home/trb/team-comtress-2/game/bin/vphysics.so error=/home/trb/team-comtress-2/game/bin/vphysics.so: undefined symbol: __wrap_freopen
 failed to dlopen vphysics.so error=/home/trb/team-comtress-2/game/bin/vphysics.so: undefined symbol: __wrap_freopen
Failed to load vphysics.so: (null)
AppFramework : Unable to load module vphysics.so!
Unable to load interface VPhysics031 from vphysics.so
Process 62744 exited with status = 0 (0x00000000)

**Comment by melvyn2** [comment]
I also got past the DXmath segfault with a dumb hack, how did you solve it? That problem and a couple afterwards I have queued to push later.

**Comment by thereal-bumface** [comment]
I rebuilt it? ./build.sh -dd, ran into same errors with ledit and lmalloc from before. Had to fix the ldl2 link that was missing as well. Weird! I'm very excited now, this is cool we're making progress.

**Comment by melvyn2** [comment]
Pull recent commits, and install the packages it asks. Then try again

**Comment by melvyn2** [comment]
The fact that it started working on your droplet makes me think that maybe this is a regression in gcc, I'll test a clang build and see where I get.

**Comment by thereal-bumface** [comment]
It'll be a bit; rebuilding `./build.sh -dd`

**Comment by thereal-bumface** [comment]
Let it build overnight, got this when I try to run it:

```
trb@tc2:~/team-comtress-2$ ./run.sh -s -d
(lldb) target create "/home/trb/team-comtress-2/game/srcds_linux"
Current executable set to '/home/trb/team-comtress-2/game/srcds_linux' (i386).
(lldb) settings set -- target.run-args  "-allowdebug" "-steam" "-game" "tf" "-insecure" "-novid" "-nojoy" "-nosteamcontroller" "-nohltv" "-particles" "1" "-noborder" "-particle_fallback" "2" "-dev" "-nobreakpad" "-console"
(lldb) r
Process 112563 launched: '/home/trb/team-comtress-2/game/srcds_linux' (i386)

WARNING: Failed to load 32-bit libtinfo.so.5 or libncurses.so.5.
  Please install (lib32tinfo5 / ncurses-libs.i686 / equivalent) to enable readline.

System (VMaterialSystem081) failed during stage CONNECTION
Process 112563 exited with status = 0 (0x00000000)
(lldb) bt
error: invalid thread
(lldb)
```

**Comment by thereal-bumface** [comment]
Hrm, got the seg fault when using the original vector4d.h. Not sure what was causing the above issue with your latest version.

**Comment by melvyn2** [comment]
the matsystem has been a bug with the dedicated server since the original port

**Comment by thereal-bumface** [comment]
Gotcha; ran again with your version (no changes) and got the same as above issue, failed during stage CONNECTION. Not sure where to go from here.

**Comment by melvyn2** [comment]
Hm, this does seem like a compiler bug, on my debian docker instance compiling in debug mode doesn't cause the vector4d crash. It wouldn't be the first time for tc2

**Comment by thereal-bumface** [comment]
Yeah, I've been running this on a Digital Ocean Droplet now, so it's no longer on Docker, getting the CONNECTION issue. Backtrace doesn't reveal anything immediate.

**Comment by thereal-bumface** [comment]
Not sure if this helps, but when I attempt to build with clang, I get the following:
```
trb@tc2:~/team-comtress-2$ ./build.sh -l
trb@tc2:~/team-comtress-2
```

**Comment by thereal-bumface** [comment]
Also, when I attempt to link:
```
cp: cannot stat '/mnt/volume_sfo3_01/TC2/tf/gamestate.txt': No such file or directory
cp: cannot stat '/mnt/volume_sfo3_01/TC2/tf/glshaders.cfg': No such file or directory
cp: cannot stat '/mnt/volume_sfo3_01/TC2/tf/videoconfig_linux.cfg': No such file or directory
```

**Comment by thereal-bumface** [comment]
Any ideas on where to look into next to fix this issue? I'm pretty much stuck, and you seem to know this project way better than myself.

**Comment by melvyn2** [comment]
when I have time I will. There’s a reason the pr is still a draft

**Comment by thereal-bumface** [comment]
Absolutely, I appreciate the time you've spent so far. Cheers!

**Comment by thereal-bumface** [comment]
Hey melvyn2, don't want to bug you too much; you get a chance to take a look at this any further? It's ok if not.

**Comment by ChampionCynthia** [comment]
Hello,

I'm currently working on getting Team Comtress 2 running on the Mac ([repo](https://github.com/ChampionCynthia/team-comtress-2-osx/tree/osx-build)). The Source Engine version of the Mac is built with Clang.

Since Team Comtress 2 has introduced C++17 stuff in the codebase, the code cannot be built using Xcode 6.4 (based on an unspecified SVN revision of LLVM 3.6.0 during its development); this version of LLVM doesn't support C++17. Thus, I did some changes to get the code to build in Xcode 9.4.1, which is the last version of Xcode to support building for i386 architectures.

I "fixed" the SSE issue from `vector4d.h` by replacing the file with [CS:GO's](https://github.com/perilouswithadollarsign/cstrike15_src/blob/master/public/mathlib/vector4d.h). The CS:GO code base usually has better code for most things, but compatibility must be checked for since the CS:GO engine is also quite different.
EDIT: It might break models however, so changes may be needed.

`srcds_osx` has the same problem as `srcds_linux`'s above:

```
System (VMaterialSystem081) failed during stage CONNECTION
2022-07-31 13:15:30.657339+0200 srcds_osx[38373:147142] 227:  ca_debug_string: conv == NULL
Process 38373 exited with status = 0 (0x00000000) 
```

Could this be caused by `shaderapiempty.dylib`? By default, we're not building that library, we're simply using TF2's.

EDIT: `hl2_osx` boots and can load a map, although without sound. But models are not positioned correctly and dynamic models are mostly broken. This may be caused by the `vector4d.h` change though.
